### PR TITLE
references #1, added disk store abstraction and used it in segmentati…

### PR DESCRIPTION
**Pull request by prashanthellina**
…on and diarization

When I run the segmentation or speaker diarization pipelines on very long audio files (>= 2 hours) the existing code is taking up too much RAM and my process is killed by the OOM killer.

In this PR I introduce a `DiskStore` abstraction which I employed in many places in the segmentation and diarization pipelines to back some numpy arrays and/or Pytorch tensors with disk storage (using memory mapping).

After the modifications, I am able to successfully run the pipeline on my long audio files (with significantly reduced memory consumption and minimal impact on the run time).

Using Disk Store:
```python
from pyannote.audio.utils.diskstore import DiskStore

dstore_path = "/home/prashanth/tmp/dstore"
if not os.path.exists(dstore_path):
    os.makedirs(dstore_path)
dstore = DiskStore(dstore_path)

...

diarization = pipeline(audio, hook=hook, disk_store=dstore)
dstore.cleanup() # deletes on disk files
```

I noticed that loading the Audio input itself was a substantial cause of memory consumption so I wrote this piece of code to alleviate the memory load. It wasn't obvious where and how to integrate the functionality into pyannote so I am pasting the code below. In addition to conserving memory, it also decodes the input audio using multiple cores to speed up the operation for large files. Note that if the sample rate of the the input audio file is not 16Khz as expected by the pipelines, I use `ffmpeg` to convert first (in order to prevent the existing code from downsampling and creating an in-memory copy of the audio data)

```python

from deeputil import generate_random_string
from soundfile import SoundFile
import ffmpeg
import numpy as np
import torch

import multiprocessing
import os
import concurrent
from concurrent.futures.process import ProcessPoolExecutor

NCORES = 4

fpath = "..." # path to very long audio file

def get_ranges(n, n_parts=0, n_per_part=0, offset=0):
    assert n_parts or n_per_part
    assert not (n_parts and n_per_part)

    if n_parts:
        n_per_part = int(n / n_parts)
    else:
        n_parts = int(n / n_per_part)

    n_excess = n % n_per_part
    parts = zip(range(0, n - 1, n_per_part), [n_per_part] * n_parts)
    parts = [[ps + offset, pd] for ps, pd in parts]
    parts[-1][-1] += n_excess

    return parts


def decode_audio(inp_fpath, out, start, nframes, nframes_per_part=48000 * 5):
    out = get_waveform("r+")
    with SoundFile(inp_fpath) as inp:
        ranges = get_ranges(nframes, n_per_part=nframes_per_part, offset=start)
        for part_start, part_frames in ranges:
            inp.seek(part_start)
            inp.read(part_frames, out=out[0][part_start : part_start + part_frames])


samplerate = 0
nframes = 0
with SoundFile(fpath) as inp:
    nframes = inp.frames
    samplerate = inp.samplerate

if samplerate != model.hparams.sample_rate:
    tmp_fpath = os.path.join(
        "/home/prashanth/tmp/dstore/audio_%s.flac" % generate_random_string(length=10)
    )
    istream = ffmpeg.input(fpath)
    ostream = istream.output(
        tmp_fpath, acodec="flac", ar=model.hparams.sample_rate, format="flac"
    )
    ffmpeg.run(ostream, quiet=True)
    fpath = tmp_fpath

with SoundFile(fpath) as inp:
    nframes = inp.frames
    samplerate = inp.samplerate

parts = get_ranges(nframes, n_parts=NCORES)

waveform_fpath = os.path.join(
    "/home/prashanth/tmp/dstore/audio_%s" % generate_random_string(length=10)
)

def get_waveform(mode):
    return np.memmap(waveform_fpath, mode=mode, shape=(1, nframes), dtype=np.float32)


waveform = get_waveform("w+")

futures = []
with ProcessPoolExecutor(max_workers=NCORES) as executor:
    for pstart, pframes in parts:
        fut = executor.submit(decode_audio, fpath, None, pstart, pframes)
        futures.append(fut)

    futures, _ = concurrent.futures.wait(futures)
    results = [fut.result() for fut in futures]

waveform = torch.from_numpy(waveform)
audio = dict(waveform=waveform, sample_rate=samplerate)
```

I noticed these issues on this repo which might be related to this
- https://github.com/pyannote/pyannote-audio/issues/1201
- https://github.com/pyannote/pyannote-audio/issues/1165
- https://github.com/pyannote/pyannote-audio/issues/1156
- https://github.com/pyannote/pyannote-audio/issues/1090

I am still working on figuring out a way to get past memory consumption issues with Agglomerative Clustering on a very large number of embeddings. If I am able to make progress there, I will issue a separate PR.

### CODE REVIEW

File: pyannote/audio/core/inference.py [hunk 1]
```diff
@@ -458,8 +484,23 @@ def aggregate(
...
+
+            np.isnan(scores.data, out=np_is_nan)
+            np.subtract(1, np_is_nan, out=masks)
+
+            _s = scores.sliding_window
+            masks = SlidingWindowFeature(
+                masks, SlidingWindow(start=_s.start, duration=_s.duration, step=_s.step)
+            )
+
+        scores.data = np.nan_to_num(scores.data, copy=False, nan=0.0)
```
Comment (prashanthellina) [inline pyannote/audio/core/inference.py:503]: I disabled making a copy here as it did not see necessary.

File: pyannote/audio/utils/signal.py [hunk 2]
```diff
@@ -123,22 +137,86 @@ def binarize_ndarray(
...
+
+
+@jit(nopython=True)
+def where_with_out(inp, x, y, out):
+    ni, nj = inp.shape
+    for i in range(ni):
+        for j in range(nj):
+            out[i, j] = x[i, j] if inp[i, j] else y[i, j]
+
+    return out
```
Comment (prashanthellina) [inline pyannote/audio/utils/signal.py:219]: I had to introduce `numba` as a dependency as I could not find a `numpy` mechanism to do fancy indexing and `where` without making a copy.

**Comment by prashanthellina** [comment]
@hbredin Do you have any thoughts on this PR? Is there anything else I can provide to get this merged?

**Comment by hbredin** [comment]
Thanks for this PR that aims at solving a problem that seems to be recurrent to many people.

However, I'd rather first focus on simpler things first, like:

* avoiding copy when unnecessary (like you did [here](https://github.com/pyannote/pyannote-audio/pull/1216#discussion_r1061776124))
* switching from `np.float32` to `np.float16` dtype for raw segmentation (or even smaller precision if that exists)
* switching from `np.float32` to `np.bool` dtype for binarized segmentation
* deleting arrays as soon as they are no longer needed
* avoiding duplication (like [this](https://github.com/pyannote/pyannote-audio/blob/6927ced52de68b6be978d3417acadee4d4ca5bf2/pyannote/audio/pipelines/speaker_diarization.py#L494) and [that](https://github.com/pyannote/pyannote-audio/blob/6927ced52de68b6be978d3417acadee4d4ca5bf2/pyannote/audio/pipelines/utils/diarization.py#L139))

But, first things first, one should run a memory profiler first to really understand where things go nuts in terms of memory. I just don't have the bandwidth to work on this right now.

Would you work on that?

**Comment by prashanthellina** [comment]
Thanks for your thoughts @hbredin . I will try running `memray` (I remember you mentioning that in a different thread) and see what is going on first.

**Comment by mflage** [comment]
@prashanthellina I see you closed this issue. Is this solved? Is there a way we can use the PR to analyze large files as well?

**Comment by nicolasfischoeder** [comment]
@prashanthellina Can you please comment on that? Is it fixed now? Is your disk store abstraction approach merged into the current development stack now?

**Comment by prashanthellina** [comment]
@nicolasfischoeder After @hbredin's feedback in the same thread, I kept `DiskStore` based changes here - https://github.com/prashanthellina/pyannote-audio.

When I find some time in the near future, I intend to profile memory usage and see if there are simpler approaches to conserving memory.

If you are interested, you can still use my changes from the repo linked here. For my current use-case I am using it and it works well. Note that it is good to use the memory mapped audio loading code from the snippet in the description of this pull request. Let me know if I can help in any way.

**Comment by nicolasfischoeder** [comment]
Hey @prashanthellina Thank you for your response. Can you please tell me how to use the hook=hook parm in the pipeline constructor?

**Comment by prashanthellina** [comment]
```python

def hook(event, *args, **kwargs):
    print(event, args,  kwargs)

diarization = pipeline(audio, hook=hook, disk_store=dstore)
```

The `hook` argument here is a `callable` that is invoked at various stages during the diarization progress and can used to print/log progress. In my application, I use it for logging and for driving a progress bar using `tqdm`.

**Comment by mflage** [comment]
> @nicolasfischoeder @mflage After @hbredin's feedback in the same thread, I kept `DiskStore` based changes here - https://github.com/prashanthellina/pyannote-audio.
> 
> When I find some time in the near future, I intend to profile memory usage and see if there are simpler approaches to conserving memory.
> 
> If you are interested, you can still use my changes from the repo linked here. For my current use-case I am using it and it works well. Note that it is good to use the memory mapped audio loading code from the snippet in the description of this pull request. Let me know if I can help in any way.

How can I actually use your implementation? I want to perform this task on a 9 hour audio clip and I find that my OOM-killer kills the process after a few seconds, even though I have 30 GB RAM and a P40 Tesla. Is there some other way of loading a large audio file? What is the limiting factor here? What is actually loaded into RAM?

**Comment by prashanthellina** [comment]
When I tested in 2 hour to 4 hour long audio files recorded at 48Khz, I noticed

1. Memory is taken up during the initial load of the audio file (all the uncompressed audio frames)
2. Because the model expects 16khz audio, a copy is made of is audio which also occupies memory.
3. During segmentation, there are some datastructures that are held in memory (relatively small)
4. Diarization takes up a lot of memory for some datastructures (I don't remember where exactly off the top of my head but I handled that in my version of the code).
5. Clustering also occupies memory when handed the speaker embeddings (I still haven't gotten around to handling this part).

For preventing the audio input file from using up memory, I posted a snippet of code that converts the audio file (downsamples) to 16khz if required and provides  this "audio = dict(waveform=waveform, sample_rate=samplerate)" . This `audio` can then be handed to the model. As the numpy array representing the audio is memory mapped, you won't hit any memory limits no matter how big the file is (it is processed in small chunks).

Once you have loaded the audio file like above, you can do

```python
from pyannote.audio.utils.diskstore import DiskStore

dstore_path = "/home/prashanth/tmp/dstore"
if not os.path.exists(dstore_path):
    os.makedirs(dstore_path)
dstore = DiskStore(dstore_path)

...

diarization = pipeline(audio, hook=hook, disk_store=dstore)
dstore.cleanup() # deletes on disk files
```

The `DiskStore` here offloads many of the memory intensive numpy arrays to the disk by memory mapping them. I am able to process a 2 hour long audio file in about 6-8GB of RAM.

File: pyannote/audio/core/inference.py (hunk 1 above)
Comment (Arnold134777) [reply pyannote/audio/core/inference.py:514]: I encountered the same problem, troubled for many days, debug the code, change here to solved.

**Pull request closed without merging**